sigmastate/src/main/scala/sigmastate/interpreter/PrecompiledScriptProcessor.scala

```diff
@@ -0,0 +1,266 @@
+package sigmastate.interpreter
+
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.atomic.AtomicInteger
+
+import com.google.common.cache.{CacheBuilder, RemovalNotification, RemovalListener, LoadingCache, CacheLoader, CacheStats}
+import org.ergoplatform.settings.ErgoAlgos
+import org.ergoplatform.validation.SigmaValidationSettings
+import org.ergoplatform.validation.ValidationRules.{CheckCostFunc, CheckCalcFunc, trySoftForkable}
+import scalan.{AVHashMap, Nullable}
+import sigmastate.Values
+import sigmastate.Values.ErgoTree
+import sigmastate.eval.{RuntimeIRContext, IRContext}
+import sigmastate.interpreter.Interpreter.{ReductionResult, WhenSoftForkReductionResult}
+import sigmastate.serialization.ErgoTreeSerializer
+import sigmastate.utils.Helpers._
+import spire.syntax.all.cfor
+
+import scala.collection.mutable
+
+/** A reducer which represents precompiled script reduction function.
+  * The function takes script execution context and produces the [[ReductionResult]],
+  * which contains both sigma proposition and the approximation of the cost taken by the
+  * reduction.
+  */
+trait ScriptReducer {
+  /** Reduce this pre-compiled script in the given context.
+    * This is equivalent to reduceToCrypto, except that graph construction is
+    * completely avoided.
+    */
+  def reduce(context: InterpreterContext): ReductionResult
+}
+
+/** Used as a fallback reducer when precompilation failed due to soft-fork condition. */
+case object WhenSoftForkReducer extends ScriptReducer {
+  override def reduce(context: InterpreterContext): (Values.SigmaBoolean, Long) = {
+    WhenSoftForkReductionResult
+  }
+}
+
+/** This class implements optimized reduction of the given pre-compiled script.
+  * Pre-compilation of the necessary graphs is performed as part of constructor and
+  * the graphs are stored in the given IR instance.
+  *
+  * The code make the following assumptions:
+  * 1) the given script doesn't contain both [[sigmastate.utxo.DeserializeContext]] and
+  * [[sigmastate.utxo.DeserializeRegister]]
+  *
+  * The code should correspond to reduceToCrypto method, but some operations may be
+  * optimized due to assumptions above.
+  */
+case class PrecompiledScriptReducer(scriptBytes: Seq[Byte])(implicit val IR: IRContext)
+  extends ScriptReducer {
+
+  /** The following operations create [[RCostingResultEx]] structure for the given
```

Comment:
Invalid ScalaDoc link, please add type annotation for the public member

Reply:
fixed